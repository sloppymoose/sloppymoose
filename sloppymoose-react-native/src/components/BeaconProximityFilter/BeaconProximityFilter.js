import { arrayOf, number, oneOf, shape, string, func } from 'prop-types'
import emptyAry from 'empty/array'
import { find, uniq } from 'lodash'
import React from 'react'

const PROXIMITY = {
  far: 2,
  immediate: 0,
  near: 1,
  unknown: 99
}

function byProximity (beaconA, beaconB) {
  if (PROXIMITY[beaconA.proximity] < PROXIMITY[beaconB.proximity]) {
    return -1
  }
  if (PROXIMITY[beaconA.proximity] > PROXIMITY[beaconB.proximity]) {
    return 1
  }
  if (beaconA.accuracy < beaconB.accuracy) {
    return -1
  }
  if (beaconA.accuracy > beaconB.accuracy) {
    return 1
  }
  return 0
}

export default class BeaconProximityFilter extends React.Component {
  static propTypes = {
    children: func.isRequired,
    rangedBeacons: arrayOf(
      shape({
        accuracy: number,
        proximity: oneOf(['far', 'immediate', 'near']),
        rssi: number,
        uuid: string.isRequired
      }).isRequired
    ),
    rangedRegions: arrayOf(
      shape({
        identifier: string.isRequired,
        major: number,
        minor: number,
        uuid: string.isRequired
      }).isRequired
    )
  }
  state = {
    nearestBeacons: emptyAry,
    nearestRegions: emptyAry
  }
  componentWillMount () {
    this.sortByProximity(this.props)
  }
  componentWillReceiveProps (nextProps) {
    // Only need to checked for changed beacons because regions will not change
    // without a change in beacons
    if (this.props.rangedBeacons !== nextProps.rangedBeacons) {
      this.sortByProximity(nextProps)
    }
  }
  sortByProximity = ({ rangedBeacons, rangedRegions }) => {
    const nearestBeacons = rangedBeacons.sort(byProximity)
    const nearestRegions = uniq(nearestBeacons.map(b => b.uuid)).map(uuid =>
      find(rangedRegions, { uuid })
    )
    this.setState({ nearestBeacons, nearestRegions })
  }
  render () {
    return this.props.children(this.state)
  }
}
