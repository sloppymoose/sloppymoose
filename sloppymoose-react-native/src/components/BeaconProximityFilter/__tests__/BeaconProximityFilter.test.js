import BeaconProximityFilter from '../BeaconProximityFilter'
import React from 'react'
import { FAR, NEAR } from './__fixtures__/beacons'
import { REGION_A, REGION_B } from './__fixtures__/regions'
import renderer from 'react-test-renderer'

describe('BeaconProximityFilter', () => {
  describe('#componentWillMount', () => {
    it('sorts regions by proximity', done => {
      renderer.create(
        <BeaconProximityFilter
          rangedBeacons={[...FAR, ...NEAR]}
          rangedRegions={[REGION_A, REGION_B]}
        >
          {({ nearestBeacons, nearestRegions }) => {
            expect(nearestBeacons).toMatchSnapshot()
            expect(nearestRegions).toMatchSnapshot()
            done()
            return <i />
          }}
        </BeaconProximityFilter>
      )
    })
  })

  describe('#componentWillReceiveProps', () => {
    describe('with new ranged beacon regions', () => {
      it('re-sorts the ranged beacon and regions', () => {
        const subject = renderer.create(
          <BeaconProximityFilter
            rangedBeacons={NEAR}
            rangedRegions={[REGION_A]}
          >
            {() => <i />}
          </BeaconProximityFilter>
        )
        const instance = subject.getInstance()
        instance.componentWillReceiveProps({
          rangedBeacons: [...FAR, ...NEAR],
          rangedRegions: [REGION_A, REGION_B]
        })
        expect(instance.state).toMatchSnapshot()
      })
    })
  })
})
