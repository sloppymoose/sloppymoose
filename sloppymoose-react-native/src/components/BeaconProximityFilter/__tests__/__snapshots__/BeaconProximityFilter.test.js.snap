// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeaconProximityFilter #componentWillMount sorts regions by proximity 1`] = `
Array [
  Object {
    "accuracy": 0.01,
    "major": 1,
    "minor": 2,
    "proximity": "near",
    "rssi": -54,
    "uuid": "REGION_B_1_1",
  },
  Object {
    "accuracy": 0.5,
    "major": 1,
    "minor": 1,
    "proximity": "near",
    "rssi": -54,
    "uuid": "REGION_B_1_1",
  },
  Object {
    "accuracy": 1,
    "major": 1,
    "minor": 2,
    "proximity": "far",
    "rssi": -54,
    "uuid": "REGION_A_1_1",
  },
  Object {
    "accuracy": 2,
    "major": 1,
    "minor": 1,
    "proximity": "far",
    "rssi": -54,
    "uuid": "REGION_A_1_1",
  },
]
`;

exports[`BeaconProximityFilter #componentWillMount sorts regions by proximity 2`] = `
Array [
  Object {
    "identifier": "REGION B",
    "major": 1,
    "minor": 1,
    "uuid": "REGION_B_1_1",
  },
  Object {
    "identifier": "REGION A",
    "major": 1,
    "minor": 1,
    "uuid": "REGION_A_1_1",
  },
]
`;

exports[`BeaconProximityFilter #componentWillReceiveProps with new ranged beacon regions re-sorts the ranged beacon and regions 1`] = `
Object {
  "nearestBeacons": Array [
    Object {
      "accuracy": 0.01,
      "major": 1,
      "minor": 2,
      "proximity": "near",
      "rssi": -54,
      "uuid": "REGION_B_1_1",
    },
    Object {
      "accuracy": 0.5,
      "major": 1,
      "minor": 1,
      "proximity": "near",
      "rssi": -54,
      "uuid": "REGION_B_1_1",
    },
    Object {
      "accuracy": 1,
      "major": 1,
      "minor": 2,
      "proximity": "far",
      "rssi": -54,
      "uuid": "REGION_A_1_1",
    },
    Object {
      "accuracy": 2,
      "major": 1,
      "minor": 1,
      "proximity": "far",
      "rssi": -54,
      "uuid": "REGION_A_1_1",
    },
  ],
  "nearestRegions": Array [
    Object {
      "identifier": "REGION B",
      "major": 1,
      "minor": 1,
      "uuid": "REGION_B_1_1",
    },
    Object {
      "identifier": "REGION A",
      "major": 1,
      "minor": 1,
      "uuid": "REGION_A_1_1",
    },
  ],
}
`;
